src/uct/ib/rc/accel/rc_mlx5.inl

```diff
@@ -1567,19 +1593,77 @@ uct_rc_mlx5_iface_common_copy_to_dm(uct_rc_mlx5_dm_copy_data_t *cache, size_t hd
     }
 }
 
-static ucs_status_t UCS_F_ALWAYS_INLINE
-uct_rc_mlx5_common_dm_make_data(uct_rc_mlx5_iface_common_t *iface,
-                                uct_rc_mlx5_dm_copy_data_t *cache,
-                                size_t hdr_len, const void *payload,
-                                unsigned length,
-                                uct_rc_iface_send_desc_t **desc_p,
-                                void **buffer_p, uct_ib_log_sge_t *log_sge)
+static void UCS_F_ALWAYS_INLINE uct_rc_mlx5_iface_copy_iov_to_dm(
+        uct_rc_mlx5_dm_copy_data_t *cache, size_t hdr_len, const uct_iov_t *iov,
+        size_t iovcnt, void *dm, uct_ib_log_sge_t *log_sge)
+{
+    typedef uint64_t misaligned_t UCS_V_ALIGNED(1);
+
+    ucs_iov_iter_t iov_iter;
+    uint64_t word;
+    size_t to_copy;
+
+    ucs_assert(cache != NULL);
+    ucs_assert(hdr_len != 0);
+    ucs_assert(sizeof(cache->bytes) >= hdr_len);
+    ucs_iov_iter_init(&iov_iter);
+
+    /* copy head of iov to tail of cache */
+    uct_iov_to_buffer(iov, iovcnt, &iov_iter,
+                      UCS_PTR_BYTE_OFFSET(cache->bytes, hdr_len),
+                      sizeof(cache->bytes) - hdr_len);
+
+    UCS_STATIC_ASSERT(sizeof(*cache) == sizeof(cache->bytes));
+    UCS_STATIC_ASSERT(ucs_static_array_size(log_sge->sg_list) >= 2);
+
+    /* atomically by 8 bytes copy data to DM */
+    /* cache buffer must be aligned, so, source data type is aligned */
+    UCS_WORD_COPY(volatile uint64_t, dm, uint64_t, cache->bytes,
+                  sizeof(cache->bytes));
+    dm = UCS_PTR_BYTE_OFFSET(dm, sizeof(cache->bytes));
+    if (ucs_log_is_enabled(UCS_LOG_LEVEL_TRACE_DATA)) {
+        log_sge->sg_list[0].addr   = (uint64_t)cache;
+        log_sge->sg_list[0].length = (uint32_t)hdr_len;
+
+        log_sge->sg_list[1].addr   = (uint64_t)iov->buffer;
+        log_sge->sg_list[1].length = (uint32_t)iov->length;
+
+        log_sge->num_sge = 2;
+    } else {
+        log_sge->num_sge = 0;
+    }
+
+    while (iov_iter.iov_index < iovcnt) {
+        to_copy = ucs_align_down(
+                iov[iov_iter.iov_index].length - iov_iter.buffer_offset,
+                sizeof(word));
+        UCS_WORD_COPY(volatile uint64_t, dm, misaligned_t,
+                      UCS_PTR_BYTE_OFFSET(iov[iov_iter.iov_index].buffer,
+                                          iov_iter.buffer_offset),
+                      to_copy);
+        dm                      = UCS_PTR_BYTE_OFFSET(dm, to_copy);
+        iov_iter.buffer_offset += to_copy;
+
+        word = 0;
+        uct_iov_to_buffer(iov, iovcnt, &iov_iter, &word, sizeof(word));
+        UCS_WORD_COPY(volatile uint64_t, dm, uint64_t, &word, sizeof(word));
+        dm   = UCS_PTR_BYTE_OFFSET(dm, sizeof(word));
```

Comment:
maybe do this part only if  iov_iter.buffer_offset < iov[iov_iter.iov_index].length ?
and add a comment like "copy the remainder of the buffer along with few bytes from the next buffer"